src/value_param.rs

```diff
@@ -0,0 +1,203 @@
+    #[doc(hidden)]
+    fn get_ptr(&mut self) -> *mut T;
+}
+
+impl<T> ValueParam<T> for &T
+where
+    T: CopyNew,
+{
+    type StackStorage = T;
+
+    fn needs_stack_space(&self) -> bool {
+        true
+    }
+
+    fn populate_stack_space(&self, this: Pin<&mut MaybeUninit<Self::StackStorage>>) {
+        unsafe { crate::moveit::new::copy(*self).new(this) }
+    }
+
+    fn get_ptr(&mut self) -> *mut T {
+        unreachable!()
+    }
+}
+
+impl<T> ValueParam<T> for UniquePtr<T>
+where
+    T: UniquePtrTarget,
+{
+    type StackStorage = ();
+
+    fn needs_stack_space(&self) -> bool {
+        false
+    }
+
+    fn populate_stack_space(&self, _: Pin<&mut MaybeUninit<Self::StackStorage>>) {}
+
+    fn get_ptr(&mut self) -> *mut T {
+        (unsafe {
+            Pin::into_inner_unchecked(
+                self.as_mut()
+                    .expect("Passed a NULL UniquePtr as a C++ value parameter"),
+            )
+        }) as *mut T
+    }
+}
+
+impl<T> ValueParam<T> for &UniquePtr<T>
+where
+    T: UniquePtrTarget + CopyNew,
+{
+    type StackStorage = T;
+
+    fn needs_stack_space(&self) -> bool {
+        true
+    }
+
+    fn populate_stack_space(&self, this: Pin<&mut MaybeUninit<Self::StackStorage>>) {
+        // Invoke a copy constructor on the Rust side, because we'll use std::move on the C++
+        // side.
+        unsafe {
+            crate::moveit::new::copy(
+                self.as_ref()
+                    .expect("Passed a NULL &UniquePtr as a C++ value parameter"),
+            )
+            .new(this)
+        }
+    }
+
+    fn get_ptr(&mut self) -> *mut T {
+        unreachable!()
+    }
+}
+
+/// Implementation detail for how we pass value parameters into C++.
+/// This type is instantiated by auto-generated autocxx code each time we
+/// need to pass a value parameter into C++, and will take responsibility
+/// for extracting that value parameter from the [`ValueParam`] and doing
+/// any later cleanup.
+#[doc(hidden)]
+pub struct ValueParamHandler<T, VP: ValueParam<T>> {
+    param: VP,
+    space: Option<MaybeUninit<VP::StackStorage>>,
+    _pinned: PhantomPinned,
+}
+
+impl<T, VP: ValueParam<T>> ValueParamHandler<T, VP> {
+    /// Create a new storage space for something that's about to be passed
+    /// by value to C++. Depending on the `ValueParam` type passed in,
+    /// this may be largely a no-op or it may involve storing a whole
+    /// extra copy of the type.
+    ///
+    /// # Safety
+    ///
+    /// Callers must guarantee that this type will not move
+    /// in memory.
+    pub unsafe fn new(param: VP) -> Self {
+        let mut this = Self {
+            param,
+            space: None,
+            _pinned: PhantomPinned,
+        };
+        if this.param.needs_stack_space() {
+            this.space = Some(MaybeUninit::uninit());
+            this.param
+                .populate_stack_space(Pin::new_unchecked(this.space.as_mut().unwrap()));
```

Comment:
What prevents this Rust object (which is the storage for a C++ object) from being memcpyed to a different address? Some of the tests fail for me after this PR, I think because of this.

With some additional prints, it looks like the `drop` for `ValueParamHandler` is definitely being called with a different address.

Reply:
Hmm, thanks, I will revert or fix this tomorrow.

Reply:
I was suspicious of my pinning logic here and that's why I had raised #833 on myself! I plan to disable this code in #852 for now, and then reinstate it as I do #847 or #821, either of which will require this `ValueParam` code to be rejigged a bit.